### PR TITLE
chore: upgrade typescript to 4.8, improve compile performance

### PR DESCRIPTION
**Pull request by thantos**
Fixes type issues which broke in TS 4.8.

Resolved a complex type that was causing an extra second of processing.

### Before
Total time:                  16.91s

```ts
    if (this.parent === undefined) {
      return undefined;
    } else if (is(this.parent)) {
      return this.parent;
    } else {
      return this.parent.findParent(is);
    }
```

### After
Total time:                  16.03s

```ts
    if (this.parent === undefined) {
      return undefined;
    } else if (is(this.parent)) {
      return this.parent as unknown as N;
    } else {
      return this.parent.findParent(is);
    }
```

Evaluating `Parent & {} & N` was taking a second

```
Hot Spots
├─ Check file /home/sussmans/functionless/src/node.ts (1741ms)
│  └─ Compare types 7794 and 1985 (1125ms)
│     └─ Compare types 7794 and 9 (1125ms)
│        ├─ {"id":7794,"kind":"Intersection","count":3,"types":[838,33,1979]}
│        │  ├─ {"id":838,"kind":"TypeParameter","name":"Parent","location":{"path":"/home/sussmans/functionless/src/node.ts","line":76,"char":3}}
│        │  ├─ {"id":33,"kind":"Other","display":"{}"}
│        │  └─ {"id":1979,"kind":"TypeParameter","name":"N","location":{"path":"/home/sussmans/functionless/src/node.ts","line":182,"char":21}}
│        └─ {"id":9,"kind":"Intrinsic","name":"undefined"}
└─ Check file /home/sussmans/functionless/src/asl/synth.ts (1267ms)
```

### CODE REVIEW

**Comment by thantos** [comment]
woops, didn't mean to grab the website change too, fixing...

**Review by sam-goodwin — APPROVED** [review]

**Comment by sam-goodwin** [comment]
Fix the pr title?

File: src/expression.ts [hunk 1]
```diff
@@ -815,7 +818,10 @@ export class TemplateSpan<
   }
 }
 
-export class TemplateMiddle extends BaseNode<NodeKind.TemplateMiddle> {
+export class TemplateMiddle extends BaseNode<
+  NodeKind.TemplateMiddle,
+  TemplateSpan
+> {
```
Comment (sam-goodwin) [inline src/expression.ts:824]: Default types don't work anymore?

File: src/node.ts [hunk 2]
```diff
@@ -57,7 +56,6 @@ export type FunctionlessNode =
   | Err
   | BindingPattern
   | ImportKeyword
-  | NoSubstitutionTemplateLiteralExpr
```
Comment (sam-goodwin) [inline src/node.ts:60]: Why did you remove this

File: src/expression.ts (hunk 1 above)
Comment (thantos) [reply src/expression.ts:824]: Trying to reduce the complexity by specifying types were possible. TemplateMiddle can only be in a TemplateSpan

File: src/node.ts (hunk 2 above)
Comment (thantos) [reply src/node.ts:60]: Its also in `Expr`.

File: src/expression.ts (hunk 1 above)
Comment (sam-goodwin) [reply src/expression.ts:824]: Oh right - good point. This is more precise.
